agentspawner/daemon.py

```diff
@@ -0,0 +1,155 @@
+        self.tickets = []
+        self.conn = resalloc_connection
+        self.log = logger
+
+    def call_converge_to(self):
+        """ Execute the configured hook script """
+        while True:
+            result = subprocess.run(["./hook-converge-to"], capture_output=True,
+                                    check=False)
+            if result.returncode == 0:
+                try:
+                    return int(result.stdout.decode("utf-8").strip())
+                except ValueError:
+                    pass
+
+            self.log.debug("Failing to run converge-to hook")
+
+    def call_take(self, data):
+        """
+        Call hook that prepares the resource
+        """
+        return not subprocess.run(["./hook-take", f"{data}"], check=True)
+
+    def call_release(self, data):
+        """
+        Call hook that releases the resource
+        """
+        result = subprocess.run(["./hook-release", f"{data}"], check=False)
+        return not result.returncode
+
+    def call_throwaway(self, data):
+        """
+        Check if agent can not be used anymore.
+        """
+        result = subprocess.run(["./hook-throwaway", f"{data}"], check=False)
+        return not result.returncode
+
+    def start(self, count):
+        """ Start N agent-like resources """
+        self.log.info("Starting %s resources", count)
+        for _ in range(count):
+            ticket = self.conn.newTicket(self.tags)
+            self.log.debug("Taking ticket id %s", ticket.id)
+            self.tickets.append(ticket.id)
+            data = ticket.wait()
+            self.call_take(data)
+
+    def _drop_ticket(self, ticket):
+        self.log.debug("Closing ticket %s", ticket.id)
+        ticket.close()
+        self.tickets.remove(ticket.id)
+
+
+    def try_to_stop(self, to_stop):
+        """
+        Attempt to stop TO_STOP resources by closing Resalloc tickets.  Not all
+        the resources may be closed at this time.
+        """
+        self.log.info("Trying to stop %s resources", to_stop)
+        stopped = 0
+        for ticket_id in list(self.tickets):
+            if stopped >= to_stop:
+                break
+
+            ticket = Ticket(ticket_id, connection=self.conn)
+            data = ticket.collect()
+            if not self.call_release(data):
+                self.log.debug("Can't release %s", ticket.id)
+                continue
+
+            self._drop_ticket(ticket)
+            stopped += 1
+
+    def recyle(self):
+        """
+        If the agent isn't able to terminate itself, we can implement a periodic
+        check here and terminate it from the outside.
+        """
+        self.log.info("Searching staled tickets")
+
+        for ticket_id in list(self.tickets):
+            ticket = Ticket(ticket_id, connection=self.conn)
+            data = ticket.collect()
+            if not self.call_throwaway(data):
+                continue
+
+            self._drop_ticket(ticket)
```

Comment:
Shall we call the release hook here? Shall the `throwaway` hook only check if a node can be thrown away or delete it too? That may cause duplication of code.

Reply:
I don't think we want to call release hook on two places. If something needs to be done on both places, then you can appropriately tweak both the configurable *release* and *throwaway* hooks.

Reply:
What is the difference between `release` and `throwaway` hooks? They seem to be doing the same thing under different names.

Reply:
It should be actually named `try_release` vs. `trhowaway`, or something like that.

The first one removes even workers who have not yet started any tasks.  While `throwaway` removes workers that already finished at least one task.